### PR TITLE
Address many Rawtypes warnings

### PR DESCRIPTION
**Pull request by madrob**
I tried to split this into multiple commits after the fact, no guarantees that they will compile individually but they are meant to be reasonably logical groupings for review.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/api/AnnotatedApi.java [hunk 1]
```diff
@@ -101,14 +101,14 @@ public EndPoint getEndPoint() {
    *                then absence of Api-s is silently ignored.
    * @return list of discovered Api-s
    */
-  public static List<Api> getApis(Class<? extends Object> theClass , Object obj, boolean allowEmpty)  {
+  public static List<Api> getApis(Class<?> theClass , Object obj, boolean allowEmpty)  {
```
Comment (fdalsotto) [inline solr/core/src/java/org/apache/solr/api/AnnotatedApi.java:104]: would an interface and Class<Interface> not have better typing than Class<?>?

File: solr/core/src/java/org/apache/solr/client/solrj/embedded/EmbeddedSolrServer.java [hunk 2]
```diff
@@ -253,7 +252,9 @@ public void writeResults(ResultContext ctx, JavaBinCodec codec) throws IOExcepti
             createJavaBinCodec(callback, resolver).setWritableDocFields(resolver).marshal(rsp.getValues(), out);
 
             try (InputStream in = out.toInputStream()) {
-              return (NamedList<Object>) new JavaBinCodec(resolver).unmarshal(in);
+              @SuppressWarnings({"unchecked"})
+              NamedList<Object> resolved = (NamedList<Object>) new JavaBinCodec(resolver).unmarshal(in);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/client/solrj/embedded/EmbeddedSolrServer.java:256]: does a one-liner no longer work as it was?

File: solr/test-framework/src/java/org/apache/solr/cloud/ZkTestServer.java [hunk 3]
```diff
@@ -644,8 +644,10 @@ public void shutdown() throws IOException, InterruptedException {
 
       while (true) {
         try {
-          zooThread.join();
-          ObjectReleaseTracker.release(zooThread);
+          if (zooThread != null) {
```
Comment (dsmiley) [inline solr/test-framework/src/java/org/apache/solr/cloud/ZkTestServer.java:647]: belongs in another issue?

File: solr/core/src/java/org/apache/solr/handler/StreamHandler.java [hunk 4]
```diff
@@ -141,15 +141,21 @@ public static void addExpressiblePlugins(StreamFactory streamFactory, SolrCore c
...
   public static class ExpressibleHolder extends PackagePluginHolder<Class<? extends Expressible>> {
+    private Class<? extends Expressible> clazz;
+
     public ExpressibleHolder(PluginInfo info, SolrCore core, SolrConfig.SolrPluginInfo pluginMeta) {
       super(info, core, pluginMeta);
     }
 
-    // WARNING: This is called from the super constructor, do not make assumptions about object state.
-    // This will probably cause a really bad bug someday, I hope you are reading this under better circumstances.
+    @Override
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/handler/StreamHandler.java:150]: The commit "potential class loading issue" is not related to the PR

**Review by dsmiley — COMMENTED** [review]
Several commits here are not about Rawtypes warnings and thus should be separated.

File: solr/core/src/java/org/apache/solr/api/AnnotatedApi.java (hunk 1 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/api/AnnotatedApi.java:104]: Probably, but that’s a bigger design decision

File: solr/core/src/java/org/apache/solr/client/solrj/embedded/EmbeddedSolrServer.java (hunk 2 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/client/solrj/embedded/EmbeddedSolrServer.java:256]: Can’t suppress the warning on a return statement.

File: solr/test-framework/src/java/org/apache/solr/cloud/ZkTestServer.java (hunk 3 above)
Comment (madrob) [reply solr/test-framework/src/java/org/apache/solr/cloud/ZkTestServer.java:647]: This came up while I was trying to debug a failing test, but sure

File: solr/core/src/java/org/apache/solr/handler/StreamHandler.java (hunk 4 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/handler/StreamHandler.java:150]: It is, because it was a bug I introduced in this issue.

**Comment by madrob** [comment]
> Several commits here are not about Rawtypes warnings and thus should be separated.

unchecked casts go hand in hand with raw types so I’m not splitting those out. The zoo thread bull check can be left out, sure. Is there more?

**Comment by dsmiley** [comment]
> unchecked casts go hand in hand with raw types so I’m not splitting those out.

100% agree; I didn't mean to suggest otherwise!

An example: TestDistribPackageStore.assertResponseValues felt out of place.  Maybe there are others.  I was expecting this PR just to be about types/casts to remove some warnings.  Thanks for this!

**Comment by madrob** [comment]
That’s a bug I found because I changed an unchecked makeMap to map.of somewhere, so iteration order changed and I discovered that we were only asserting the first value in the map.

Sometimes when you remove warnings you find real bugs! If this wasn’t the case, I wouldn’t spend time on it.

File: solr/core/src/java/org/apache/solr/core/SolrConfig.java [hunk 5]
```diff
@@ -622,15 +622,22 @@ public HttpCachingConfig getHttpCachingConfig() {
 
   public static class HttpCachingConfig implements MapSerializable {
 
+    /**
+     * config xpath prefix for getting HTTP Caching options
+     */
+    private static final String CACHE_PRE
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/SolrConfig.java:628]: I can't tell where this is used

**Review by dsmiley — COMMENTED** [review]
Wow; tons of work here in those first two commits!  I'm glad you are tending to this long task.

**Comment by dsmiley** [comment]
It'll be nice to see this changes merged... did you forget about this one @madrob?

**Comment by madrob** [comment]
There was a merge conflict with some of Noble’s work on the XPath stuff so I wanted to wait for a merge of that, but I suppose there’s no major reason to hold up. I’ll take care of this on Tuesday

File: solr/core/src/java/org/apache/solr/core/SolrConfig.java (hunk 5 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/core/SolrConfig.java:549]: https://github.com/apache/solr/pull/107/files#diff-473fbcdab103c08461ad1b3c3bb1c6d56f1bcd16d6ce341d80855db2cb20a427R591

**Comment by dsmiley** [comment]
I used IntelliJ's "inspections" to find cases where we no longer need to suppress certain warnings.  There were false positives from precommit that I reverted back.  

ClusterStateProvider.getClusterProperty was a curious case, where the generics there were not helpful and forced the need to suppress on implementing classes even if they don't override that method.  So I removed generics there.

**Comment by madrob** [comment]
What is the IntelliJ inspection called? I tried to find it but couldn’t. :(

**Comment by dsmiley** [comment]
"Redundant suppression".  AFAICT you can't restrict to particular suppressions, and be aware it's decisions may disagree with Gradle and thus we must adhere to Gradle.  Many rawtypes suppressions were recommended to be removed yet they couldn't be because -- https://www.oreilly.com/library/view/learning-java-4th/9781449372477/ch08s10.html

**Review by dsmiley — APPROVED** [review]

**Pull request merged**